File: 4/4b.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	var grid []string

	for scanner.Scan() {
		grid = append(grid, strings.TrimSpace(scanner.Text()))
	}

	n, m := len(grid), len(grid[0])
	words := map[string]bool{"MAS": true, "SAM": true}
	ans := 0

	for i := 0; i < n-2; i++ {
		for j := 0; j < m-2; j++ {
			left := string(grid[i][j]) + string(grid[i+1][j+1]) + string(grid[i+2][j+2])
			right := string(grid[i][j+2]) + string(grid[i+1][j+1]) + string(grid[i+2][j])
			if words[left] && words[right] {
				ans++
			}
		}
	}

	fmt.Println(ans)
}

File: 4/4b.py
import sys

grid = [line.strip() for line in sys.stdin]

n, m = len(grid), len(grid[0])

ans = 0

words = ['MAS','SAM']

for i in range(n - 2):
    for j in range(m - 2):
        left = f"{grid[i][j]}{grid[i + 1][j + 1]}{grid[i + 2][j + 2]}"
        right = f"{grid[i][j + 2]}{grid[i + 1][j + 1]}{grid[i + 2][j]}"
        ans += left in words and right in words

print(ans)




        